activesupport/lib/active_support/core_ext/time/calculations.rb

```diff
@@ -47,7 +47,9 @@ def at_with_coercion(*args)
       # Time.at can be called with a time or numerical value
       time_or_number = args.first
 
-      if time_or_number.is_a?(ActiveSupport::TimeWithZone) || time_or_number.is_a?(DateTime)
+      if time_or_number.is_a?(ActiveSupport::TimeWithZone)
+        at_without_coercion(time_or_number.to_r).getlocal
```

Comment:
This is the alternative implementation I'm pushing for in https://github.com/rails/rails/pull/40137. https://github.com/rails/rails/issues/40413 is a duplicate of https://github.com/rails/rails/issues/38831, although demonstrated slightly differently.

I'd like to wait a bit and see if https://github.com/rails/rails/pull/40137 gets merged, since it was opened first and the author clearly invested a lot of time into it already.